nx/lib/nx/binary_backend.ex

```diff
@@ -1571,6 +1571,89 @@ defmodule Nx.BinaryBackend do
     end
   end
 
+  @impl true
+  def scatter_add(
+        %T{} = out,
+        %T{shape: shape, type: {_, target_size}} = target,
+        %T{shape: {indices_rows, _indices_cols} = indices_shape} = indices,
+        %T{shape: {indices_rows}} = updates,
+        _opts
+      ) do
+    indices_bin_list =
+      indices |> to_binary() |> aggregate_axes([1], indices_shape, elem(indices.type, 1))
+
+    target_byte_size = div(target_size, 8)
+
+    offsets_list =
+      match_types [indices.type] do
+        for idx_bin <- indices_bin_list do
+          idx = for <<match!(x, 0) <- idx_bin>>, do: read!(x, 0)
+
+          {offset, []} =
+            idx
+            |> Enum.with_index()
+            |> Enum.reduce(
+              {0, Tuple.to_list(shape)},
+              fn {idx, axis}, {offset, [dim_size | shape]} ->
+                if idx < 0 or idx >= dim_size do
+                  raise ArgumentError,
+                        "index #{idx} is out of bounds for axis #{axis} in shape #{inspect(shape)}"
+                end
+
+                {offset + idx * Enum.product(shape), shape}
+              end
+            )
+
+          offset * target_byte_size
+        end
+      end
+
+    updates_list =
+      match_types [updates.type] do
+        for <<match!(x, 0) <- to_binary(updates)>>, do: read!(x, 0)
+      end
+
+    {offsets_with_updates, _last_offset} =
+      offsets_list
+      |> Enum.zip(updates_list)
+      |> Enum.group_by(fn {off, _} -> off end, fn {_, upd} -> upd end)
+      |> Enum.sort_by(fn {off, _} -> off end)
+      |> Enum.map_reduce(0, fn {next_offset, upds}, previous_offset ->
+        {{
+           previous_offset + target_byte_size,
+           next_offset,
+           Enum.sum(upds)
+         }, next_offset}
+      end)
+
+    target_binary = to_binary(target)
+
+    {result, tail} =
+      for {previous, current, update} <- offsets_with_updates, reduce: {<<>>, target_binary} do
+        {traversed, to_traverse} ->
+          before_slice_size = max(current - previous, 0)
+          before_offset = binary_part(to_traverse, 0, before_slice_size)
```

Comment:
@josevalim are all our types 8-byte aligned? I've just realized that this line should be followed by a match-read-write combo to cast from one type to another... The updated elements are being treated properly.

Reply:
Well, anyway, I fixed this in the latest commit, although I couldn't reproduce any problems with f32 + s64 or f32 + f64

Reply:
They are _8-bit_ aligned, yes. Btw, instead of binary_part, consider using the binary syntax I showed, it is going to be more concise and more efficient:

    <<pre::size(pre), match!(...), post::binary>>

:)

Reply:
I was facing some issues with this syntax, but I'll work into a refactor for this now :)

Reply:
https://github.com/elixir-nx/nx/pull/464